source/beamline/BioXAS/BioXASValvesControl.cpp

```diff
@@ -0,0 +1,421 @@
+#include "BioXASValvesControl.h"
+#include "actions3/AMListAction3.h"
+#include "actions3/AMActionSupport.h"
+
+BioXASValvesControl::BioXASValvesControl(const QString &name, QObject *parent) :
+	AMPseudoMotorControl(name, "", parent)
+{
+	// Initialize inherited variables.
+
+	value_ = None;
+	setpoint_ = None;
+	minimumValue_ = None;
+	maximumValue_ = Closed;
+
+	setEnumStates(QStringList() << valueToString(None) << valueToString(Open) << valueToString(Closed));
+	setMoveEnumStates(QStringList() << valueToString(Open));
+
+	setContextKnownDescription("ValvesControl");
+
+	// Create member variables.
+
+	frontEndValveSet_ = 0;
+	sideValveSet_ = 0;
+	mainValveSet_ = 0;
+	imagingValveSet_ = 0;
+
+	// Current settings.
+
+	updateStates();
+}
+
+BioXASValvesControl::~BioXASValvesControl()
+{
+
+}
+
+bool BioXASValvesControl::canMeasure() const
+{
+	return false;
+}
+
+bool BioXASValvesControl::canMove() const
+{
+	return false;
+}
+
+bool BioXASValvesControl::validValue(double value) const
+{
+	bool result = false;
+
+	switch (int(value)) {
+	case None:
+		result = true;
+		break;
+	case Open:
+		result = true;
+		break;
+	case Closed:
+		result = true;
+		break;
+	default:
+		break;
+	}
+
+	return result;
+}
+
+bool BioXASValvesControl::validSetpoint(double value) const
+{
+	bool result = false;
+
+	// Only want to be able to use this control to view current
+	// state, and to open valves (not close them).
+
+	if (value == BioXASValvesControl::Open)
+		result = true;
+
+	return result;
+}
+
+AMControlSet* BioXASValvesControl::valveSet(BioXAS::Beamline beamline)
+{
+	AMControlSet *result = 0;
+
+	switch (beamline) {
+	case BioXAS::FrontEnd:
+		result = frontEndValveSet_;
+		break;
+	case BioXAS::Side:
+		result = sideValveSet_;
+		break;
+	case BioXAS::Main:
+		result = mainValveSet_;
+		break;
+	case BioXAS::Imaging:
+		result = imagingValveSet_;
+		break;
+	default:
+		break;
+	}
+
+	return result;
+}
+
+bool BioXASValvesControl::isOpen() const
```

Comment:
Yup, the confirmed requirement is: no beamline is operational when any vacuum valve on any beamline is closed. In this respect, the beamlines act like a unit and I've been assured that there will never (never!) arise a situation where a valve would be closed on one beamline and another would still be running. To get this to work, the pvs themselves would have to be circumvented; Acquaman wouldn't know the difference either way.